File: JavaScript/ds/Trie.js
TrieNode = function (key, parent) {
  this.key = key;
  this.count = 0;
  this.children = Object.create(null);
  if (parent === undefined) {
    this.parent = null;
  } else {
    this.parent = parent;
  }
};

function Trie() {
  // create only root with null key and parent
  this.root = new TrieNode(null, null);
}

// Recursively finds the occurence of all words in a given node
Trie.findAllWords = function (root, word, output) {
  if (root === null) return;
  if (root.count > 0)
    if (typeof output === "object")
      output.push({ word: word, count: root.count });
  var key;
  for (key in root.children) {
    word += key;
    this.findAllWords(root.children[key], word, output);
    word = word.slice(0, -1);
  }
};

Trie.prototype.insert = function (word) {
  if (typeof word !== "string") return;
  if (word === "") {
    this.root.count += 1;
    return;
  }
  var node = this.root;
  var len = word.length;
  var i;
  for (i = 0; i < len; i++) {
    if (node.children[word.charAt(i)] === undefined)
      node.children[word.charAt(i)] = new TrieNode(word.charAt(i), node);
    node = node.children[word.charAt(i)];
  }
  node.count += 1;
};

Trie.prototype.findPrefix = function (word) {
  if (typeof word !== "string") return null;
  var node = this.root;
  var len = word.length;
  var i;
  // After end of this loop node will be at desired prefix
  for (i = 0; i < len; i++) {
    if (node.children[word.charAt(i)] === undefined) return null; // No such prefix exists
    node = node.children[word.charAt(i)];
  }
  return node;
};

Trie.prototype.remove = function (word, count) {
  if (typeof word !== "string") return;
  if (typeof count !== "number") count = 1;
  else if (count <= 0) return;

  // for empty string just delete count of root
  if (word === "") {
    if (this.root.count >= count) this.root.count -= count;
    else this.root.count = 0;
    return;
  }

  var child = this.root;
  var len = word.length;
  var i, key;
  // child: node which is to be deleted
  for (i = 0; i < len; i++) {
    key = word.charAt(i);
    if (child.children[key] === undefined) return;
    child = child.children[key];
  }

  // Delete no of occurences specified
  if (child.count >= count) child.count -= count;
  else child.count = 0;

  // Since some occurences are left we don't delete it
  if (child.count > 0) return;
  // If some child nodes of current node exists
  else if (child.children !== Object.create(null)) return;
  // If code reaches here we can delete the node
  // for deleting node we just remove its reference from
  // parent so the GC can collect it
  else child.parent.children[child.key] = undefined;
};

Trie.prototype.findAllWords = function (prefix) {
  var output = [];
  // find the node with provided prefix
  var node = this.findPrefix(prefix);
  // No such prefix exists
  if (node === null) return output;
  Trie.findAllWords(node, prefix, output);
  return output;
};

Trie.prototype.contains = function (word) {
  // find the node with given prefix
  var node = this.findPrefix(word);
  // No such word exists
  if (node === null || node.count === 0) return false;
  return true;
};

Trie.prototype.findOccurences = function (word) {
  // find the node with given prefix
  var node = this.findPrefix(word);
  // No such word exists
  if (node === null) return 0;
  return node.count;
};

// To test
(function demo() {
  var x = new Trie();
  x.insert("sheldon");
  x.insert("hello");
  x.insert("anyword");
  x.insert("sheldoncooper");
  console.log(x.findOccurences("sheldon"));
  x.remove("anything");
  x.insert("sheldon");
  console.log(x.findOccurences("sheldon"));
  console.log(x.findAllWords("sheldon"));
  x.insert("anything");
  x.remove("sheldoncooper");
  console.log(x.contains("sheldoncooper"));
  console.log(x.findAllWords("sheldon"));
})();

File: Python/ds/Trie.py
"""
Tries is a kind of search tree data structure used to store a dynamic set or associative array where the keys are usually strings.

In this implementation we have only focused on maintaining a Trie for lowercase letter words only.

Time Complexity: O (length of longest string in Trie * Number of queries)

Space Complexity: O (size of node * total nodes)
"""

class Trienode:
	"""
	Each node in the Trie will contain the following structure
	"""
	def __init__(self,char):
		self.char=char
		self.wordend=0
		self.children=[None for i in range(26)]

class Trie:
	"""
	The Trie data structure
	"""
	def __init__(self):
		"""
		The root will contain a '/' character. We may also keep it an empty string, it won't matter.
		"""
		self.root=Trienode('/')

	def insert(self,word):
		"""
		To insert word in Trie
		"""
		ptr=self.root

		for c in word:
			idx=ord(c) - ord('a')  # This is done for relative indexing of the children array

			if ptr.children[idx] != None:
				ptr=ptr.children[idx]
				ptr.pc+=1

			else:
				ptr.children[idx]=Trienode(c)
				ptr=ptr.children[idx]
				ptr.pc+=1

		ptr.wordend = 1

	def _search(self,word):
		"""
		Searches if a word is present in the Trie. Basically a private function that can be used to query or delete results
		"""
		ptr=self.root
		self.exists=True

		for c in word:
			idx = ord(c)-ord('a')

			if ptr.children[idx] != None:
				ptr=ptr.children[idx]

			# Note: In the following lines of code, specific numbers and a pointer is returned just for the convinience to map 
			# the types of results returned from the Trie. We may use other numbers or approach as well. 

			else:
				# not exists as ptr has no child in that idx, i.e, it's value is None
				self.exists=False
				return (-1,None)

		if self.exists:

			if ptr.wordend == 1:
				# exists
				return (0,ptr)
			else:
				# not exist as wordend= 0
				return (1,None)

	def query(self,word):
		"""
		To query if a word is present in Trie
		"""
		val, _= self._search(word)

		if val==-1 or val==1:
			print(word,"does not exist in the Trie")
		elif val ==0:
			print(word,"exists in the Trie")

	def delete(self,word):
		"""
		To delete a word from Trie
		"""
		val,ptr = self._search(word)

		if val== -1:
			print(word,'does not exists in Trie')
		
		elif val== 1 :
			print(word,'does not exists in Trie')

		elif val==0:
			ptr.wordend =0
			print(word,'Deleted')


def menu(trie):
	
	while True:
		print('Enter I/i: insert, Q/q: query, D/d: delete, E/e: exit')
		ip=input()

		if ip=='I' or ip=='i':
			print('Enter a word: ')
			word=input()
			trie.insert(word)

		elif ip=='Q' or ip=='q':
			print('Enter a word: ')
			word=input()
			trie.query(word)

		elif ip =='D' or ip=='d':
			print('Enter a word: ')
			word=input()
			trie.delete(word)

		elif ip =='E' or ip=='e':
			break


def main():

	"""
	#Sample Test cases
	ip=['aab','acd','bab','aa']
	s=['aab','ac','baba','ace','aa']

	trie=Trie()

	for word in init_ip:
		trie.insert(word)

	for word in s:
		trie.query(word)
	
	trie.delete('ba')
	trie.delete('bab')
	trie.delete('aaba')
	trie.delete('bab')
	"""

	#Through real time user input
	print("Input strings to be inserted in the Trie (Note: only lowercase words)")
	init_ip=input().split()

	trie=Trie()

	for word in init_ip:
		trie.insert(word)
	
	menu(trie)

if __name__=='__main__':
	main()








